### ISSUE TITLE
reconcile Future/Promise

### ISSUE THREAD

**Issue by jmesserly** [dart-lang/sdk#27315]
_From @jmesserly on June 25, 2015 20:50_

For JS interop, it would be ideal if we could adapt Dart Futures to implement the Promise interface.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

This will also ease DOM interop as Promise becomes more common.

Ideally, Dart Future actually is a JS Promise, simply exposing different user-facing APIs (similar to Dart builtin List type and JS Array). But we could also try adapters, similar to Dart Iterable objects exposing a `[Symbol.iterator]()` method to JS, and Dart for-each consuming JS Iterables.

_Copied from original issue: dart-lang/dev_compiler#245_

**Comment by jmesserly** [comment]
related #221

**Comment by jmesserly** [comment]
_From @jacob314 on January 4, 2016 19:10_

This is now becoming an issue for JS interop

**Comment by jmesserly** [comment]
yeah, just wanted to add an update: this is something we're definitely doing, only question is when it happens. Probably after we've gotten the Dart language tests and modular compilation in a good place.

**Comment by jmesserly** [comment]
and SDK issue is https://github.com/dart-lang/sdk/issues/24679

**Comment by rayk** [comment]
This issue has been hanging around for almost a year.

Actually I first hit it [here](http://stackoverflow.com/questions/34572517/dart-js-interop-0-6-0-and-js-promises-resolving). From my understanding there has not been any work on this. I know that @alan-knight was playing some Service worker stuff (using all dart:html), but look at the repo it looks like it did not move forward, possibly because the register required handling the promise.

If my assumptions above are correct?

Then certainly this needs to addressed in a once and for all, promises are spreading in JS and have it map one to one with a Future in the dart2JS is reasonable. Otherwise there will be just be promise completer boilerplate all over the place.

There is a type definition for [es6 promises](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/es6-promise) yet [JS Facade Gen](https://github.com/dart-lang/js_facade_gen) does not appear to handle it too well @jacob314.

Hence I for one would like to this issue given some priority and clarity.

**Comment by jmesserly** [comment]
Hit this via https://github.com/dart-lang/sdk/issues/28422.

I'm going to see if we can do adapters in both directions:

* create a _PromiseFuture internal type that maps to native JS Promise, but implements Dart's Future.
* inject a compatible "then" and "catch" method onto Dart Futures, essentially so they implement the Promise interface.

Tricky part here is figuring out how to inject "then" and "catch" because those aren't symbolized and collide with the Dart names. Then again ... once the Future interface is marked as "may be native" we might get the renaming for free.

**Comment by daniel-v** [comment]
Would be nice to have this.

**Comment by lexaknyazev** [comment]
If/when dart compilers get native support of JS Promises, please don't tie it to `dart:html`. 

FWIW, we're using the following boilerplate which works mostly fine both in browsers and node:
```dart
@JS()
class Promise<T> {
  external Promise(void executor(void resolve(T result), Function reject));
  external Promise then(void onFulfilled(T result), [Function onRejected]);
}
```

To create a `Promise` from `Future` (to send it to JS world):
```dart
return new Promise<MyType>(allowInterop((resolve, reject) {
  myFuture.then(resolve, onError: reject);
}));
```

To create a `Future` from `Promise` (gotten from JS world):
```dart
final completer = new Completer<MyType>();
myPromise.then(allowInterop(completer.complete),
    allowInterop(completer.completeError));
return completer.future;
```

**Comment by Sunbreak** [comment]
> If/when dart compilers get native support of JS Promises, please don't tie it to `dart:html`.
> 
> FWIW, we're using the following boilerplate which works mostly fine both in browsers and node:
> 
> ```dart
> @JS()
> class Promise<T> {
>   external Promise(void executor(void resolve(T result), Function reject));
>   external Promise then(void onFulfilled(T result), [Function onRejected]);
> }
> ```
> 
> To create a `Promise` from `Future` (to send it to JS world):
> 
> ```dart
> return new Promise<MyType>(allowInterop((resolve, reject) {
>   myFuture.then(resolve, onError: reject);
> }));
> ```
> 
> To create a `Future` from `Promise` (gotten from JS world):
> 
> ```dart
> final completer = new Completer<MyType>();
> myPromise.then(allowInterop(completer.complete),
>     allowInterop(completer.completeError));
> return completer.future;
> ```

`Flutter for Web` is beta now. The `Promise` implementation seems to work in debug mode but not profile/release mode.
Could you show some light on me?

**Comment by Sunbreak** [comment]
It is promising!

https://github.com/dart-lang/sdk/blob/a577d0aea51f242e72bd8d39f8b3368c33b4a4f3/sdk/lib/_internal/js_dev_runtime/lib/js_util/dart2js/js_util_dart2js.dart#L132-L150

```dart
/// Converts a JavaScript Promise to a Dart [Future].
///
/// ```dart
/// @JS()
/// external Promise<num> get threePromise; // Resolves to 3
///
/// final Future<num> threeFuture = promiseToFuture(threePromise);
///
/// final three = await threeFuture; // == 3
/// ```
Future<T> promiseToFuture<T>(jsPromise) {
  final completer = Completer<T>();

  final success = convertDartClosureToJS((r) => completer.complete(r), 1);
  final error = convertDartClosureToJS((e) => completer.completeError(e), 1);

  JS('', '#.then(#, #)', jsPromise, success, error);
  return completer.future;
}
```

**Comment by jodinathan** [comment]
any news on this?

would be awesome to be able to await a promise without having to use ```promiseToFuture```
